RFD.md

```diff
@@ -0,0 +1,191 @@
+1. Start Job
+    - process is assigned a job ID (UUID) and added to an in-memory process map
+    - output file is created, identified by the job ID
+    - process is started as a child process in user mode under `rpsuser`
+      - process output is redirected to the output file on process start
+      - process is added to the cgroup on process start using the linux `cgexec` utility, which will allow running a process directly within a specified control group without having to manually add it
+2. Job Running
+    - process status can be queried by accessing the process map
+    - process output can be streamed by reading the output file
+3. Stop Job
+    - when process ends, the SIGCHLD handler will be invoked
+    - process map status is updated (checking status will say `stopped`)
+    - output file cleanup is scheduled for 1 hour after process end
+
+As mentioned `job_id` will be a UUID, which is unique and less likely to be reused than a PID. This will allow us to distinguish between processes that had the same PID when getting the status/output of a job.
+
+The asynchronous SIGCHLD handler will avoid blocking the main thread and will ensure that the process terminates immediately and gracefully (no zombie state) and invoke cleanup.
+
+Processes will be run by a dedicated non-sudoer user (e.g. `rpsuser`) to ensure that jobs are not run as root or sudo. This will be a different user than the API server user, which will be run as `root` to be able to manage cgroups. `rpsuser` will not have any permissions to server code or output files (more deails below)
+
+#### Output Streaming
+
+Process output will be streamed to a disk file identified by the job ID, rather than kept in memory. The tradeoffs are:
+
+| Feature | Memory stream | Disk File |
+|---|---|---|
+| Read/Write performance | Faster | Slower |
+| Resource usage | RAM (potentially high) | Disk, low RAM |
+| Concurrent read implementation | More Complex (channels, mutex, etc.) | Easier (OS handles concurrent reads) |
+| Persistence after process stop | More complex | Simpler |
+
+While memory streams offer faster read/write speeds, they require more RAM and complex synchronization for concurrent access. Disk files, though slower, use less RAM, are simpler to manage for concurrent reads (handled by the OS), and persist after process termination. To avoid excessive disk usage, output files will be cleaned up after 1 hour.
+
+_Note:_ In a production system, we would want a more sophisticated output file management, such as moving them to an external system, and automatic cleanup if disk space gets low. Keeping it simple for this demo
+
+Output files will be created by the `root` user with 600 permissions. The `rpsuser` running jobs will not have write access to any output file, in order to prevent users from messing with these files. Instead, the server will pass an open file descriptor to the child process, which will write to the output file.
```

Comment:
No need to do this `root` `rpsuser` split.

Reply:
got it, removed this so processes will run as root